draft/celery-5-high-level-architecture.rst

```diff
@@ -0,0 +1,650 @@
+    'max_retries': 3,
+    'interval_start': 0,
+    'interval_step': 0.2,
+    'interval_max': 0.2
+  })
+  def foo():
+    second_operation()
+
+  @task(retry_policy={
+    'max_retries': 10,
+    'interval_start': 0,
+    'interval_step': 5,
+    'interval_max': 120
+  })
+  def bar():
+    first_operation()
+    foo.delay()
+
+Or you can wrap each code section in a try..except clause and call
+:py:meth:`celery.app.task.Task.retry`.
+
+.. code-block:: python
+
+  @task(bind=True)
+  def foo(self):
+    try:
+      # first operation
+    except Exception:
+      self.retry(retry_policy={
+        'max_retries': 10,
+        'interval_start': 0,
+        'interval_step': 5,
+        'interval_max': 120
+      })
+
+    try:
+      first_operation()
+    except Exception:
+      self.retry(retry_policy={
+        'max_retries': 10,
+        'interval_start': 0,
+        'interval_step': 5,
+        'interval_max': 120
+      })
+
+    try:
+      second_operation()
+    except Exception:
+      self.retry(retry_policy={
+        'max_retries': 3,
+        'interval_start': 0,
+        'interval_step': 0.2,
+        'interval_max': 0.2
+      })
+
+Those solutions are unnecessarily verbose. Instead, we could use a with clause
+if all we want to do is retry.
+
+.. code-block:: python
+
+  @task
+  def foo():
+    with retry(max_retries=10, interval_start=0, interval_step=5, interval_max=120):
+      first_operation()
+
+    with retry(max_retries=10, interval_start=0, interval_step=5, interval_max=120):
+      second_operation()
+
+Health Checks
+~~~~~~~~~~~~~
+
+Circuit Breaker
+~~~~~~~~~~~~~~~
+
+Failure Resilience
+++++++++++++++++++
+
+Observability
+-------------
+
+One of Celery 5's goals is to be :term:`observable <Observability>`.
+
+Each Celery component will record statistics, provide trace points for
+application monitoring tools and distributed tracing tools and emit log messages
+when appropriate.
+
+Metrics
++++++++
+
+Trace Points
+++++++++++++
+
+Logging
++++++++
+
+All log messages must be structured.
+:term:`Structured logs <Structured Logging>` provide context for our users
+which allows them to debug problems more easily and aids the developers
+to resolve bugs in Celery.
+
+The structure of a log message is determined whenever a component
+is initialized.
+
+During initialization, an attempt will be made to detect how the component
+lifecycle is managed.
+If all attempts are unsuccessful, the logs will be formatted using
+:term:`JSON` and will be printed to stdout.
+
+Celery will provide an extension point for detection of different
+runtimes.
+
+.. admonition:: Example
+
+  If a component's lifecycle is managed by a SystemD service,
+  Celery will detect that the `JOURNAL_STREAM`_ environment variable
+  is set when the process starts and use it's value to transmit structured
+  data into `journald`_.
+
+Worker
+------
+
+Protocol
+++++++++
+
+Introduction to AMQP 1.0 Terminology
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
```

Comment:
might be an important one to consider. https://github.com/rabbitmq/rabbitmq-amqp1.0

Reply:
https://www.cloudamqp.com/blog/2018-12-10-an-update-from-the-rabbitmq-team.html